### PR TITLE
[DR-3393] Upgrade Spring 2.7.4 -> 3.1.6, TCL 0.0.75 -> 0.1.7 

### PR DESCRIPTION
**Pull request by okotsopoulos**

### CODE REVIEW

File: service/build.gradle [hunk 1]
```diff
@@ -55,32 +55,38 @@ dependencies {
...
+	implementation 'io.sentry:sentry-logback:6.19.0'
+	implementation 'org.codehaus.janino:janino:3.1.9' // Provides if/else xml parsing for logback config
 	implementation('org.springframework.cloud:spring-cloud-gcp-starter-logging:1.2.8.RELEASE') {
 		exclude group: 'org.springframework', module: 'spring-jcl'
 	}
 	implementation 'com.google.oauth-client:google-oauth-client:1.34.1'
 	implementation platform('com.google.cloud:libraries-bom:26.14.0')
 	implementation 'com.google.cloud:google-cloud-storage'
 	implementation 'io.sentry:sentry-spring-boot-starter:6.18.1'
-	implementation('org.yaml:snakeyaml') {
```
Comment (okotsopoulos) [inline service/build.gradle:67]: No longer needed:

```
(base) okotsopo@wm111-e35 terra-drs-hub % ./gradlew service:dependencyInsight --dependency snakeyaml

> Task :service:dependencyInsight
org.yaml:snakeyaml:1.33 (selected by rule)
  Variant compile:
    | Attribute Name                 | Provided | Requested    |
    |--------------------------------|----------|--------------|
    | org.gradle.status              | release  |              |
    | org.gradle.category            | library  | library      |
    | org.gradle.libraryelements     | jar      | classes      |
    | org.gradle.usage               | java-api | java-api     |
    | org.gradle.dependency.bundling |          | external     |
    | org.gradle.jvm.environment     |          | standard-jvm |
    | org.gradle.jvm.version         |          | 17           |

org.yaml:snakeyaml:1.33
\--- org.springframework.boot:spring-boot-starter:3.1.2
     +--- org.springframework.boot:spring-boot-starter-actuator:3.1.2
     |    \--- compileClasspath
     +--- org.springframework.boot:spring-boot-starter-web:3.1.2
     |    \--- compileClasspath
     +--- org.springframework.boot:spring-boot-starter-thymeleaf:3.1.2
     |    \--- compileClasspath
     +--- org.springframework.cloud:spring-cloud-gcp-starter:1.2.8.RELEASE (requested org.springframework.boot:spring-boot-starter:2.3.10.RELEASE)
     |    \--- org.springframework.cloud:spring-cloud-gcp-starter-logging:1.2.8.RELEASE
     |         \--- compileClasspath
     \--- org.springframework.boot:spring-boot-starter-json:3.1.2
          +--- org.springframework.boot:spring-boot-starter-web:3.1.2 (*)
          \--- bio.terra:externalcreds-client-resttemplate:0.72.0-SNAPSHOT:20220318.201049-1 (requested org.springframework.boot:spring-boot-starter-json:2.6.2)
               \--- compileClasspath

(*) - dependencies omitted (listed previously)
```

File: service/spring.gradle [hunk 2]
```diff
@@ -1,28 +1,22 @@
...
 
-	implementation 'org.apache.httpcomponents:httpclient:4.5.13'
+	implementation 'org.apache.httpcomponents.client5:httpclient5'
 	implementation 'org.apache.commons:commons-collections4:4.4'
 
 	// allows mocking final classes
 	testImplementation 'org.mockito:mockito-inline:5.2.0'
-
-	// See https://nvd.nist.gov/vuln/detail/CVE-2023-46589
-	// TODO DR-3393: remove this pin when upgrading Spring Boot, TCL
```
Comment (okotsopoulos) [inline service/spring.gradle:19]: No longer needed (and in fact, this earlier pin was incompatible with other upgrades):

```
(base) okotsopo@wm111-e35 terra-drs-hub % ./gradlew service:dependencyInsight --dependency tomcat-embed-core

> Task :service:dependencyInsight
org.apache.tomcat.embed:tomcat-embed-core:10.1.11 (selected by rule)
  Variant compile:
    | Attribute Name                 | Provided | Requested    |
    |--------------------------------|----------|--------------|
    | org.gradle.status              | release  |              |
    | org.gradle.category            | library  | library      |
    | org.gradle.libraryelements     | jar      | classes      |
    | org.gradle.usage               | java-api | java-api     |
    | org.gradle.dependency.bundling |          | external     |
    | org.gradle.jvm.environment     |          | standard-jvm |
    | org.gradle.jvm.version         |          | 17           |

org.apache.tomcat.embed:tomcat-embed-core:10.1.11
+--- org.apache.tomcat.embed:tomcat-embed-websocket:10.1.11
|    \--- org.springframework.boot:spring-boot-starter-tomcat:3.1.2
|         \--- org.springframework.boot:spring-boot-starter-web:3.1.2
|              \--- compileClasspath
\--- org.springframework.boot:spring-boot-starter-tomcat:3.1.2 (*)

(*) - dependencies omitted (listed previously)
```

File: service/build.gradle [hunk 3]
```diff
@@ -33,19 +33,19 @@ apply from: 'analysis.gradle'
 
 dependencyManagement {
 	dependencies {
-		dependencySet(group:'io.github.hakky54', version: '7.3.0') {
-			entry 'sslcontext-kickstart-for-apache4'
+		dependencySet(group:'io.github.hakky54', version: '8.2.0') {
```
Comment (okotsopoulos) [inline service/build.gradle:36]: This upgrade had the extra benefit of resolving https://broadworkbench.atlassian.net/browse/DR-3391 -- we needed these bouncycastle dependencies >= 1.74.

```
(base) okotsopo@wm111-e35 terra-drs-hub % ./gradlew service:dependencyInsight --dependency bouncycastle

> Task :service:dependencyInsight
org.bouncycastle:bcpkix-jdk18on:1.76
  Variant compile:
    | Attribute Name                 | Provided | Requested    |
    |--------------------------------|----------|--------------|
    | org.gradle.status              | release  |              |
    | org.gradle.category            | library  | library      |
    | org.gradle.libraryelements     | jar      | classes      |
    | org.gradle.usage               | java-api | java-api     |
    | org.gradle.dependency.bundling |          | external     |
    | org.gradle.jvm.environment     |          | standard-jvm |
    | org.gradle.jvm.version         |          | 17           |

org.bouncycastle:bcpkix-jdk18on:1.76
\--- io.github.hakky54:sslcontext-kickstart-for-pem:8.2.0
     \--- compileClasspath (requested io.github.hakky54:sslcontext-kickstart-for-pem)

org.bouncycastle:bcprov-jdk18on:1.76
  Variant compile:
    | Attribute Name                 | Provided | Requested    |
    |--------------------------------|----------|--------------|
    | org.gradle.status              | release  |              |
    | org.gradle.category            | library  | library      |
    | org.gradle.libraryelements     | jar      | classes      |
    | org.gradle.usage               | java-api | java-api     |
    | org.gradle.dependency.bundling |          | external     |
    | org.gradle.jvm.environment     |          | standard-jvm |
    | org.gradle.jvm.version         |          | 17           |

org.bouncycastle:bcprov-jdk18on:1.76
+--- org.bouncycastle:bcpkix-jdk18on:1.76
|    \--- io.github.hakky54:sslcontext-kickstart-for-pem:8.2.0
|         \--- compileClasspath (requested io.github.hakky54:sslcontext-kickstart-for-pem)
\--- org.bouncycastle:bcutil-jdk18on:1.76
     \--- org.bouncycastle:bcpkix-jdk18on:1.76 (*)

org.bouncycastle:bcutil-jdk18on:1.76
  Variant compile:
    | Attribute Name                 | Provided | Requested    |
    |--------------------------------|----------|--------------|
    | org.gradle.status              | release  |              |
    | org.gradle.category            | library  | library      |
    | org.gradle.libraryelements     | jar      | classes      |
    | org.gradle.usage               | java-api | java-api     |
    | org.gradle.dependency.bundling |          | external     |
    | org.gradle.jvm.environment     |          | standard-jvm |
    | org.gradle.jvm.version         |          | 17           |

org.bouncycastle:bcutil-jdk18on:1.76
\--- org.bouncycastle:bcpkix-jdk18on:1.76
     \--- io.github.hakky54:sslcontext-kickstart-for-pem:8.2.0
          \--- compileClasspath (requested io.github.hakky54:sslcontext-kickstart-for-pem)

(*) - dependencies omitted (listed previously)
```

File: service/spring.gradle [hunk 4]
```diff
@@ -1,28 +1,29 @@
...
-		version {
-			strictly('9.0.83')
-		}
-	}
-
+	// See https://nvd.nist.gov/vuln/detail/cve-2023-6378
+	// TODO: when latest versions of Spring are released on 2023-12-21, see if upgrading also brings
+	// its logback versions >= 1.4.14, which would allow us to stop explicitly pulling them in.
+	var logbackVersion = '1.4.14'
+	implementation "ch.qos.logback:logback-classic:$logbackVersion"
```
Comment (okotsopoulos) [inline service/spring.gradle:25]:
```
(base) okotsopo@wm111-e35 terra-drs-hub % ./gradlew service:dependencyInsight --dependency logback-classic                    

> Task :service:dependencyInsight
ch.qos.logback:logback-classic:1.4.14 (selected by rule)
  Variant compile:
    | Attribute Name                 | Provided | Requested    |
    |--------------------------------|----------|--------------|
    | org.gradle.status              | release  |              |
    | org.gradle.category            | library  | library      |
    | org.gradle.libraryelements     | jar      | classes      |
    | org.gradle.usage               | java-api | java-api     |
    | org.gradle.dependency.bundling |          | external     |
    | org.gradle.jvm.environment     |          | standard-jvm |
    | org.gradle.jvm.version         |          | 17           |

ch.qos.logback:logback-classic:1.4.14
\--- compileClasspath

ch.qos.logback:logback-classic:1.1.3 -> 1.4.14
\--- ch.qos.logback.contrib:logback-json-classic:0.1.5
     \--- org.springframework.cloud:spring-cloud-gcp-logging:1.2.8.RELEASE
          \--- org.springframework.cloud:spring-cloud-gcp-starter-logging:1.2.8.RELEASE
               \--- compileClasspath

ch.qos.logback:logback-classic:1.2.11 -> 1.4.14
\--- com.google.cloud:google-cloud-logging-logback:0.130.11-alpha
     +--- org.springframework.cloud:spring-cloud-gcp-logging:1.2.8.RELEASE (requested com.google.cloud:google-cloud-logging-logback:0.119.0-alpha)
     |    \--- org.springframework.cloud:spring-cloud-gcp-starter-logging:1.2.8.RELEASE
     |         \--- compileClasspath
     \--- com.google.cloud:libraries-bom:26.14.0
          \--- compileClasspath

ch.qos.logback:logback-classic:1.4.11 -> 1.4.14
\--- org.springframework.boot:spring-boot-starter-logging:3.1.6
     \--- org.springframework.boot:spring-boot-starter:3.1.6
          +--- org.springframework.boot:spring-boot-starter-actuator:3.1.6
          |    \--- compileClasspath
          +--- org.springframework.boot:spring-boot-starter-web:3.1.6
          |    \--- compileClasspath
          +--- org.springframework.boot:spring-boot-starter-thymeleaf:3.1.6
          |    \--- compileClasspath
          +--- org.springframework.boot:spring-boot-starter-json:3.1.6
          |    +--- org.springframework.boot:spring-boot-starter-web:3.1.6 (*)
          |    \--- bio.terra:externalcreds-client-resttemplate:0.72.0-SNAPSHOT:20220318.201049-1 (requested org.springframework.boot:spring-boot-starter-json:2.6.2)
          |         \--- compileClasspath
          \--- org.springframework.cloud:spring-cloud-gcp-starter:1.2.8.RELEASE (requested org.springframework.boot:spring-boot-starter:2.3.10.RELEASE)
               \--- org.springframework.cloud:spring-cloud-gcp-starter-logging:1.2.8.RELEASE
                    \--- compileClasspath

(*) - dependencies omitted (listed previously)
```

File: service/spring.gradle [hunk 5]
```diff
@@ -1,28 +1,29 @@
...
+	implementation "ch.qos.logback:logback-core:$logbackVersion"
```
Comment (okotsopoulos) [inline service/spring.gradle:26]:
```
(base) okotsopo@wm111-e35 terra-drs-hub % ./gradlew service:dependencyInsight --dependency logback-core   

> Task :service:dependencyInsight
ch.qos.logback:logback-core:1.4.14 (selected by rule)
  Variant compile:
    | Attribute Name                 | Provided | Requested    |
    |--------------------------------|----------|--------------|
    | org.gradle.status              | release  |              |
    | org.gradle.category            | library  | library      |
    | org.gradle.libraryelements     | jar      | classes      |
    | org.gradle.usage               | java-api | java-api     |
    | org.gradle.dependency.bundling |          | external     |
    | org.gradle.jvm.environment     |          | standard-jvm |
    | org.gradle.jvm.version         |          | 17           |

ch.qos.logback:logback-core:1.4.14
+--- compileClasspath
\--- ch.qos.logback:logback-classic:1.4.14
     +--- compileClasspath
     +--- org.springframework.boot:spring-boot-starter-logging:3.1.6 (requested ch.qos.logback:logback-classic:1.4.11)
     |    \--- org.springframework.boot:spring-boot-starter:3.1.6
     |         +--- org.springframework.boot:spring-boot-starter-actuator:3.1.6
     |         |    \--- compileClasspath
     |         +--- org.springframework.boot:spring-boot-starter-web:3.1.6
     |         |    \--- compileClasspath
     |         +--- org.springframework.boot:spring-boot-starter-thymeleaf:3.1.6
     |         |    \--- compileClasspath
     |         +--- org.springframework.boot:spring-boot-starter-json:3.1.6
     |         |    +--- org.springframework.boot:spring-boot-starter-web:3.1.6 (*)
     |         |    \--- bio.terra:externalcreds-client-resttemplate:0.72.0-SNAPSHOT:20220318.201049-1 (requested org.springframework.boot:spring-boot-starter-json:2.6.2)
     |         |         \--- compileClasspath
     |         \--- org.springframework.cloud:spring-cloud-gcp-starter:1.2.8.RELEASE (requested org.springframework.boot:spring-boot-starter:2.3.10.RELEASE)
     |              \--- org.springframework.cloud:spring-cloud-gcp-starter-logging:1.2.8.RELEASE
     |                   \--- compileClasspath
     +--- ch.qos.logback.contrib:logback-json-classic:0.1.5 (requested ch.qos.logback:logback-classic:1.1.3)
     |    \--- org.springframework.cloud:spring-cloud-gcp-logging:1.2.8.RELEASE
     |         \--- org.springframework.cloud:spring-cloud-gcp-starter-logging:1.2.8.RELEASE (*)
     \--- com.google.cloud:google-cloud-logging-logback:0.130.11-alpha (requested ch.qos.logback:logback-classic:1.2.11)
          +--- org.springframework.cloud:spring-cloud-gcp-logging:1.2.8.RELEASE (requested com.google.cloud:google-cloud-logging-logback:0.119.0-alpha) (*)
          \--- com.google.cloud:libraries-bom:26.14.0
               \--- compileClasspath

ch.qos.logback:logback-core:1.1.3 -> 1.4.14
\--- ch.qos.logback.contrib:logback-json-core:0.1.5
     \--- ch.qos.logback.contrib:logback-json-classic:0.1.5
          \--- org.springframework.cloud:spring-cloud-gcp-logging:1.2.8.RELEASE
               \--- org.springframework.cloud:spring-cloud-gcp-starter-logging:1.2.8.RELEASE
                    \--- compileClasspath

ch.qos.logback:logback-core:1.2.11 -> 1.4.14
\--- com.google.cloud:google-cloud-logging-logback:0.130.11-alpha
     +--- org.springframework.cloud:spring-cloud-gcp-logging:1.2.8.RELEASE (requested com.google.cloud:google-cloud-logging-logback:0.119.0-alpha)
     |    \--- org.springframework.cloud:spring-cloud-gcp-starter-logging:1.2.8.RELEASE
     |         \--- compileClasspath
     \--- com.google.cloud:libraries-bom:26.14.0
          \--- compileClasspath

(*) - dependencies omitted (listed previously)
```

File: build.gradle [hunk 6]
```diff
@@ -2,7 +2,7 @@ plugins {
 	id 'idea'
 
 	id 'io.spring.dependency-management' version '1.0.11.RELEASE' apply false
```
Comment (pshapiro4broad) [inline build.gradle:4]: I think this should be updated to the current spring dependency plugin version
```suggestion
	id 'io.spring.dependency-management' version '1.1.4' apply false
```

File: service/spring.gradle [hunk 7]
```diff
@@ -1,28 +1,29 @@
 configurations { compileOnly { extendsFrom annotationProcessor } }
 
-var springBootVersion = '2.7.4'
+var springBootVersion = '3.1.6'
 
 dependencies {
+	implementation "org.springframework.boot:spring-boot-starter-actuator:$springBootVersion"
```
Comment (pshapiro4broad) [inline service/spring.gradle:6]: The spring dependency plugin is supposed to eliminate the need to specify the version of spring components like spring-boot. If you update the dependency plugin to the current version, can you remove the versions from this dependency?

**Comment by okotsopoulos** [comment]
Needed to recreate this PR off of a new FB with no slashes in the name:
https://github.com/DataBiosphere/terra-drs-hub/pull/97

I want to push this branch to my Bee, but slashes in the name don't work with image publishing: https://github.com/DataBiosphere/terra-drs-hub/actions/runs/7184743682/job/19566525540

File: service/spring.gradle (hunk 7 above)
Comment (okotsopoulos) [reply service/spring.gradle:6]: Yep!

**Pull request closed without merging**